src/trio/testing/_raises_group.py

```diff
@@ -545,50 +667,199 @@ def matches(
-                    return True
-            return False
-
-        if self.match_expr is not None and not re.search(
-            self.match_expr,
-            _stringify_exception(exc_val),
-        ):
+            # we opt to only print type of the exception here, as the repr would
+            # likely be quite long
+            not_group_msg = f"{type(exc_val).__name__!r} is not an exception group"
+            if len(self.expected_exceptions) > 1:
+                self.fail_reason = not_group_msg
+                return False
+            # if we have 1 expected exception, check if it would work even if
+            # allow_unwrapped is not set
+            res = _check_expected(self.expected_exceptions[0], exc_val)
+            if res is None and self.allow_unwrapped:
+                return True
+
+            if res is None:
+                self.fail_reason = (
+                    f"{not_group_msg}, but would match with `allow_unwrapped=True`"
+                )
+            elif self.allow_unwrapped:
+                self.fail_reason = res
+            else:
+                self.fail_reason = not_group_msg
             return False
 
-        remaining_exceptions = list(self.expected_exceptions)
         actual_exceptions: Sequence[BaseException] = exc_val.exceptions
         if self.flatten_subgroups:
             actual_exceptions = self._unroll_exceptions(actual_exceptions)
 
-        # important to check the length *after* flattening subgroups
-        if len(actual_exceptions) != len(self.expected_exceptions):
+        self.fail_reason = _check_match(self.match_expr, exc_val)
+        if self.fail_reason is not None:
+            if (
+                len(actual_exceptions) == len(self.expected_exceptions) == 1
+                and isinstance(expected := self.expected_exceptions[0], type)
+                and isinstance(actual := actual_exceptions[0], expected)
+                and _check_match(self.match_expr, actual) is None
+            ):
+                assert (
+                    self.match_expr is not None
+                ), "can't be None if _check_match failed"
+                self.fail_reason += f", but matched expected exception. You might want RaisesGroup(Matcher({expected.__name__}, match={_match_pattern(self.match_expr)!r}"
             return False
 
-        for e in actual_exceptions:
-            for rem_e in remaining_exceptions:
-                if (
-                    (isinstance(rem_e, type) and isinstance(e, rem_e))
-                    or (isinstance(rem_e, RaisesGroup) and rem_e.matches(e))
-                    or (isinstance(rem_e, Matcher) and rem_e.matches(e))
-                ):
-                    remaining_exceptions.remove(rem_e)
+        # do the full check on expected exceptions
+        self.fail_reason = self._check_exceptions(
+            actual_exceptions,
+            _depth=_depth,
+        )
+
+        if self.fail_reason is not None:
```

Comment:
```suggestion
        if not self._check_exceptions(
            actual_exceptions,
            _depth=_depth,
        ):
```

Just mutate `self.fail_reason` in `self._check_exceptions`. This gets rid of the `# type: ignore` later.

Reply:
some complications:
1. The comment is incorrect, we're not passing `exc_val` to `_check_exceptions` but `actual_exceptions` which is either `exc_val.exceptions` or a flattened representation of it. So we can't do a `TypeGuard` here (at least not without doing a bunch of messy stuff) and we're gonna need a `cast` or `type: ignore` anyway.
2. A few lines below this we're rerunning `self._check_exceptions` with flattening, but ignoring `self.fail_reason`, so we'll need to juggle the old value.

Perhaps still worth it, gonna see how it looks to move the other `_check_xxx` into an abstract base class and have them set `fail_reason`